libvirt: add e2e demo code and documentation

wainersm opened the pull request:
In https://github.com/confidential-containers/cloud-api-adaptor/issues/16#issuecomment-1292524184 I drafted a framework that we could be using for running e2e tests on CI. To continue that work I want to create an implementation for libvirt because developers can run on their own workstation and this should allow others to test the framework.

But soon I realized the information to build an environment to develop for libvirt is spread in the repository or simply missing. Then I started fixing some stuffs and automating some steps. This pull request contain that work. Apart from using the changes I propose here on CI, I think it will be useful for developers setting up their environments for testing locally.

snir911 commented on libvirt/image/Makefile:
```diff
@@ -120,6 +122,19 @@ $(PAUSE_SRC): $(SKOPEO_SRC)/bin/skopeo
+		caa-libvirt-image-builder \
+		bash -c 'trap "chown -R -f "$$UID:$$GID" output staticlib $(SKOPEO_SRC) \
+		$(UMOCI_SRC) $(PAUSE_SRC) $(FILES_DIR)/$(PAUSE_BUNDLE) $(FILES_DIR)/usr \
+		../../agent-protocol-forwarder || true" EXIT; \
```
can you explain this line?

wainersm replied:
Magic my friend, magic...

Those files are created with root permission and I don't like it. For example, afterwards when I run `make clean` I get permission denied. That line ensure that `chown` will run on bash-exit and then will change the files ownership to the user that called docker. I know that there are other ways to address that issue, for example, create an user inside the container with same id; but I thought it would be too complicated. If you have a easier way to solve it then please let me know.

snir911 commented on libvirt/image/Makefile:
```diff
@@ -141,3 +144,12 @@ clean:
+push:
+ifeq ($(shell virsh -c qemu:///system vol-list --pool $(LIBVIRT_POOL) | grep $(LIBVIRT_VOL_NAME)),)
+	virsh -c qemu:///system vol-create-as --pool $(LIBVIRT_POOL) --name $(LIBVIRT_VOL_NAME) \
+		--capacity 107374182400 --allocation 2361393152 --prealloc-metadata \
```
I use `--capacity 20G --allocation 2G` for some time now and i didn't hit any issues

wainersm replied:
Cool. I will change to use those values.

snir911 commented on libvirt/kcli_cluster.sh:
```diff
@@ -0,0 +1,126 @@
+
+main() {
+	if ! command -v kcli >/dev/null; then
+		echo "ERROR: kcli command is required."
```
maybe refer to `https://kcli.readthedocs.io/en/latest/#installation` in the error?

snir911 commented on libvirt/install_operator.sh:
```diff
@@ -0,0 +1,111 @@
+script_dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+
+LIBVIRT_IP="${LIBVIRT_IP:-}"
+LIBVIRT_USER="${LIBVIRT_USER:-$USER}"
```
commit msg mentions default user is `wmoschet`, which is very cool name, it was disappointing to discover it used some other lame name on my system ;)  I'd update the commit msg so other won't be disappointed as well

wainersm replied:
lol

snir911 commented on libvirt/install_operator.sh:
```diff
@@ -0,0 +1,111 @@
+LIBVIRT_USER="${LIBVIRT_USER:-$USER}"
+LIBVIRT_NET="${LIBVIRT_NET:-default}"
+LIBVIRT_POOL="${LIBVIRT_POOL:-default}"
+SSH_KEY_FILE="${SSH_KEY_FILE:-}"
```
would it make sense to generate temporary one for the user?

wainersm replied:
Yeah, I was considering that.... for CI I will need to automatically create it. For the general case I was not sure. It seems a good idea for me. What about you?

snir911 replied:
personally i crate a new key pair rather than passing the my default private key around (which i used for many other things)

snir911 commented on Makefile:
```diff
@@ -78,5 +78,6 @@ deploy: ## Deploy cloud-api-adaptor using the operator, according to install/ove
 .PHONY: delete
 delete: ## Delete cloud-api-adaptor using the operator, according to install/overlays/$(CLOUD_PROVIDER)/kustomization.yaml file.
 	kubectl delete -k install/overlays/$(CLOUD_PROVIDER)
+	kubectl delete -f install/yamls/deploy.yaml
```
Is this required? AFAIR we deliberately didn't remove those, cc @bpradipt

wainersm replied:
hmmm... if it was deliberated then mind to explain the rationale? :)

snir911 replied:
I don't recall, probably no that important, maybe @jensfr could say if it make sense

bpradipt replied:
I think the Makefile changes should be moved to a separate PR. Reason being to discuss whether we really need to install the operator in `deploy` target ? or just refer to operator instructions ?
IOW, should the `deploy` and `delete` (is `undeploy` the right target?) only cater to caa ?

wainersm replied:
Thanks for the input @bpradipt ! I will move those makefile changes to a separate PR.

snir911 commented on libvirt/README.md:
```diff
@@ -58,3 +58,224 @@ mkdir -p /opt/data-dir
+**Note**: ensure that `~/.ssh/authorized_keys` has the right permissions (read/write for the user only) otherwise the
+authentication can silently fail.
+
+You will need to figure out the IP address of your local host (e.g. 192.168.1.107). Then try to remote connect with
```
isn't it defaults to 192.168.122.1 ? maybe because you non-root?

bpradipt replied:
default is usually 192.168.122.1. So may be the example can be changed to that.

bpradipt replied:
But it's fine as-is as well.

wainersm replied:
Done.

snir911 reviewed (commented):
thanks!

wainersm commented on libvirt/image/Dockerfile_ubuntu.builder:
```diff
@@ -0,0 +1,37 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Creates a builder container image that should be used to build the Pod VM
+# disk inside a container.
```
I played the role of fortune teller: I guess at some point developers will need to create Fedora-based Pod VM; that's why put `ubuntu` in the file name. I am okay to suffix with `podvm_builder` :)

wainersm replied:
I will change the image name from `caa-libvirt-image-builder` to `caa-libvirt-ubuntu-podvm-builder`.

wainersm commented:
Updated this PR to address the comments of @snir911. BTW, thanks for the review Snir!

wainersm commented:
Resolved the conflict on `Makefile`.

snir911 reviewed (approved):
LGTM overall thanks!

bpradipt commented on libvirt/README.md:
```diff
@@ -58,3 +58,224 @@ mkdir -p /opt/data-dir
+# Creating an end-to-end environment for testing and development
+
+In this section you will learn how to setup an environment in your local machine to run peer pods with
+the libvirt cloud API adaptor. Bear in mind that many different tools can be used to setup the environment
```
I think it's better to mention that this requires a Linux/KVM system with libvirt. Some devs do use Mac ;)

bpradipt commented on libvirt/image/Makefile:
```diff
@@ -120,9 +125,33 @@ $(PAUSE_SRC): $(SKOPEO_SRC)/bin/skopeo
 .PHONY: force
 force:
+
+.PHONY: push
```
@jtumber-ibm can you check this PR and suggest your views so that it helps with  https://github.com/confidential-containers/cloud-api-adaptor/issues/314 ?

jtumber-ibm replied:
> @jtumber-ibm can you check this PR and suggest your views so that it helps with #314 ?

The file matches up closely to the other providers and so will be simple to migrate to the common approach. The plan for #314 is to introduce the "common" Makefile first and allow the provider SMEs to update and test each provider Makefile. As this file shares the same "common" sections I'm happy for it to go in like this before #314 is ready.

wainersm replied:
Cool, I wasn't aware of #314 and I got just lucky because I copied the `push` idea from the ibmcloud's Makefile :)

wainersm commented:
Updated this PR to remove the commits with changes to the makefiles.

stevenhorsman commented on libvirt/kcli_cluster.sh:
```diff
@@ -0,0 +1,126 @@
+		wait_time=$((wait_time-"$sleep_time"))
+	done
+
+	[ "$wait_time" -gt 0 ]
```
```suggestion
	[ "$wait_time" -ge 0 ]
```
I think this should be greater than of equal to, otherwise if the process passes in the last time slot then wait_time will be 0, so 0 -gt 0 will return 'false'

wainersm replied:
good catch!

stevenhorsman commented on libvirt/kcli_cluster.sh:
```diff
@@ -0,0 +1,126 @@
+
+	local cmd="kubectl get nodes | grep '\<Ready\>.*worker'"
+	echo "Wait at least one worker be Ready"
+	if ! wait_for_process "240" "30" "$cmd"; then
```
```suggestion
	if ! wait_for_process "300" "30" "$cmd"; then
```
The worker took 274s for me, so I think increasing the timeout is needed.

stevenhorsman reviewed (commented):
Sorry Wainer, I haven't had time to finish this review as I had lots of timeouts and problems with the cluster creation, but this is what I have so far. I'll try and do more next week!

wainersm commented on libvirt/README.md:
```diff
@@ -58,3 +58,224 @@ mkdir -p /opt/data-dir
+
+- docker (or podman-docker)
+- [kcli](https://kcli.readthedocs.io/en/latest/)
+
```
That's a good idea @stevenhorsman ! I will come up with a separate PR; and on this one I will list kubectl as a requirement.

stevenhorsman commented on libvirt/image/Makefile:
```diff
@@ -120,9 +125,32 @@ $(PAUSE_SRC): $(SKOPEO_SRC)/bin/skopeo
+		bash -c 'trap "chown -R -f "$$UID:$$GID" output staticlib $(SKOPEO_SRC) \
+		$(UMOCI_SRC) $(PAUSE_SRC) $(FILES_DIR)/$(PAUSE_BUNDLE) $(FILES_DIR)/usr \
+		../../agent-protocol-forwarder || true" EXIT; \
+		source ~/.cargo/env && CLOUD_PROVIDER=libvirt make build'
```
I hit some error running this line as part of `make docker-build`:
```
docker run --rm \
	-v /root/go/src/github.com/wainersm/cc-cloud-api-adaptor:/src/cloud-api-adaptor \
	-v /root/go/src/github.com/wainersm/kata-containers:/src/kata-containers \
	-e UID=0 \
	-e GID=0 \
	caa-libvirt-ubuntu-podvm-builder \
	bash -c 'trap "chown -R -f "$UID:$GID" output staticlib skopeo \
	umoci pause files/pause_bundle files/usr \
	../../agent-protocol-forwarder || true" EXIT; \
	source ~/.cargo/env && CLOUD_PROVIDER=libvirt make build'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  5394  100  5394    0     0  54484      0 --:--:-- --:--:-- --:--:-- 54484
/bin/sh: 1: virsh: not found
```
Is that something to worry about?

wainersm replied:
It is the `ifdef` in the `push` target that is evaluated always. I re-wrote the target to fix it.

stevenhorsman replied:
No virsh error anymore. Thanks!

stevenhorsman commented on libvirt/image/Makefile:
```diff
@@ -120,9 +125,32 @@ $(PAUSE_SRC): $(SKOPEO_SRC)/bin/skopeo
+		bash -c 'trap "chown -R -f "$$UID:$$GID" output staticlib $(SKOPEO_SRC) \
+		$(UMOCI_SRC) $(PAUSE_SRC) $(FILES_DIR)/$(PAUSE_BUNDLE) $(FILES_DIR)/usr \
+		../../agent-protocol-forwarder || true" EXIT; \
+		source ~/.cargo/env && CLOUD_PROVIDER=libvirt make build'
```
I also hit a packer error here:
```
packer build -var qemu_image_name=podvm-4e38611-dirty-amd64.qcow2 \
	-var cloud_image_url=https://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-amd64.img \
	-var cloud_image_checksum=b6d04d7ed478a083547eebf009228dc72afade79fcbc097f95226512056bbc2e .
qemu.ubuntu: output will be in this color.

==> qemu.ubuntu: Retrieving ISO
==> qemu.ubuntu: Trying https://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-amd64.img
==> qemu.ubuntu: Trying https://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-amd64.img?checksum=sha256%3Ab6d04d7ed478a083547eebf009228dc72afade79fcbc097f95226512056bbc2e
==> qemu.ubuntu: https://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-amd64.img?checksum=sha256%3Ab6d04d7ed478a083547eebf009228dc72afade79fcbc097f95226512056bbc2e => /root/.cache/packer/36780b83a751001e19d6763c24d4845672af5f33.iso
==> qemu.ubuntu: Copying hard drive...
==> qemu.ubuntu: Resizing hard drive...
==> qemu.ubuntu: Found port for communicator (SSH, WinRM, etc): 2257.
==> qemu.ubuntu: Looking for available port between 5900 and 6000 on 127.0.0.1
==> qemu.ubuntu: Starting VM, booting disk image
    qemu.ubuntu: The VM will be run headless, without a GUI. If you want to
    qemu.ubuntu: view the screen of the VM, connect via VNC without a password to
    qemu.ubuntu: vnc://127.0.0.1:5965
==> qemu.ubuntu: Overriding default Qemu arguments with qemuargs template option...
==> qemu.ubuntu: Waiting 10s for boot...
==> qemu.ubuntu: Connecting to VM via VNC (127.0.0.1:5965)
==> qemu.ubuntu: Typing the boot command over VNC...
    qemu.ubuntu: Not using a NetBridge -- skipping StepWaitGuestAddress
==> qemu.ubuntu: Using SSH communicator to connect: 127.0.0.1
==> qemu.ubuntu: Waiting for SSH to become available...
==> qemu.ubuntu: Error waiting for SSH: Packer experienced an authentication error when trying to connect via SSH. This can happen if your username/password are wrong. You may want to double-check your credentials as part of your debugging process. original error: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
==> qemu.ubuntu: Deleting output directory...
Build 'qemu.ubuntu' errored after 4 minutes 50 seconds: Packer experienced an authentication error when trying to connect via SSH. This can happen if your username/password are wrong. You may want to double-check your credentials as part of your debugging process. original error: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain

==> Wait completed after 4 minutes 50 seconds

==> Some builds didn't complete successfully and had errors:
--> qemu.ubuntu: Packer experienced an authentication error when trying to connect via SSH. This can happen if your username/password are wrong. You may want to double-check your credentials as part of your debugging process. original error: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain

==> Builds finished but no artifacts were created.
make: *** [Makefile:84: podvm-4e38611-dirty-amd64.qcow2] Error 1
make: *** [Makefile:130: docker-build] Error 2
```
It might be user error, but as I think it is inside the container I'm not sure how to debug it?

wainersm replied:
Packer start qemu inside the container. It seems not able to connect to the VM via ssh after a while. In `qemu-ubuntu.pkr.hcl` it set `ssh_wait_timeout` to `300s`. Could you increase the timeout and give it a try?

stevenhorsman replied:
So I tried to double the timeout to 10mins:
```
diff --git a/libvirt/image/qemu-ubuntu.pkr.hcl b/libvirt/image/qemu-ubuntu.pkr.hcl
index 083d6ad..12e1564 100644
--- a/libvirt/image/qemu-ubuntu.pkr.hcl
+++ b/libvirt/image/qemu-ubuntu.pkr.hcl
@@ -12,7 +12,7 @@ source "qemu" "ubuntu" {
   ssh_password      = "${var.ssh_password}"
   ssh_port          = 22
   ssh_username      = "${var.ssh_username}"
-  ssh_wait_timeout  = "300s"
+  ssh_wait_timeout  = "600s"
   vm_name           = "${var.qemu_image_name}"
   shutdown_command  = "sudo shutdown -h now"
 }
```

But it still failed just short of 5 again:
```
==> qemu.ubuntu: Waiting for SSH to become available...
==> qemu.ubuntu: Error waiting for SSH: Packer experienced an authentication error when trying to connect via SSH. This can happen if your username/password are wrong. You may want to double-check your credentials as part of your debugging process. original error: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
==> qemu.ubuntu: Deleting output directory...
Build 'qemu.ubuntu' errored after 4 minutes 57 seconds: Packer experienced an authentication error when trying to connect via SSH. This can happen if your username/password are wrong. You may want to double-check your credentials as part of your debugging process. original error: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
```

So I don't know if that was enough. Sorry for being slow here - I'm exercising the path of a user who has no clue what they are doing 😉

stevenhorsman reviewed (commented):
Next round of comments/issues

wainersm commented:
Updated the PR to address @stevenhorsman 's comments.

snir911 commented on libvirt/README.md:
```diff
@@ -58,3 +58,229 @@ mkdir -p /opt/data-dir
+                                 Dload  Upload   Total   Spent    Left  Speed
+100  5394  100  5394    0     0   6421      0 --:--:-- --:--:-- --:--:--  6421
+virsh -c qemu:///system vol-create-as --pool default --name podvm-base.qcow2 \
+	--capacity 107374182400 --allocation 2361393152 --prealloc-metadata \
```
nit: change it to match the` --capacity 20G --allocation 2G `change

wainersm replied:
done

bpradipt commented:
@wainersm I have reused the builder image concept and made it generic for all providers in https://github.com/confidential-containers/cloud-api-adaptor/pull/396
Added your SoB to the commit.

snir911 commented:
FWIW,  i think created cluster worker node needs to be labeled as worker manually

stevenhorsman commented:
@wainersm, @bpradipt - Can I just check the status of this PR. I think this is the foundations of the e2e framework piece, so we want to push it forward, but I think this PR is also out of date considering the work that Pradipta has already done on with the podvm_builder, so does it need rebasing/reworking based on that?

I know I hit some errors when I tested this earlier, so is there anything we can do to help with this, or shall we just leave it until the rebase occurs? Thanks!

bpradipt commented:
@stevenhorsman we'll need to wait for @wainersm to be back. 
Currently, the podvm image can be created using docker/podman. Also if required existing base qcow2 image can be used if there are no changes to the agent and podvm files to speed up deployment and test case execution.
Given these building blocks, this PR will need changes.

wainersm commented:
hi @stevenhorsman @bpradipt , I am working on the required changes for this PR. Almost there :)

wainersm commented:
I just rebased the PR and:
 * dropped the commit that introduces the build inside container because it was moved to the podvm directory
 * dropped the commit that introduces a `make push` command to publish the qcow2 to libvirt. I will resume that work on another PR if I find worth
 * adjusted the README changes

wainersm commented:
@stevenhorsman @snir911 @bpradipt Can I get your approval to merge this?

stevenhorsman commented:
> Can I get your approval to merge this?
I'll try and get time to give this a run through today, or (more likely) tomorrow. I know last time I hit some issues, which are probably due to my lack of experience with libvirt, so apologies in advance for peppering you with questions!

bpradipt commented on libvirt/README.md:
````diff
@@ -44,3 +44,222 @@ If you want to set default password for podvm debugging then you can use guestfi
+$ virsh vol-upload --vol podvm-base.qcow2 $IMAGE --pool default --sparse
+```
+
+You should see that the `podvm-base.qcow2` volume was proper created:
````
nit: proper->properly

wainersm replied:
done

bpradipt reviewed (approved):
/lgtm
Thanks @wainersm

stevenhorsman commented on libvirt/kcli_cluster.sh:
```diff
@@ -0,0 +1,126 @@
+	kcli create kube generic \
+		-P domain="kata.com" \
+		-P pool="$LIBVIRT_POOL" \
+		-P masters="$CLUSTER_MASTERS" \
```
In my version of kcli (version: 99.0 commit: 8bb9ccd 2023/01/19 Available Updates: False), I got:
```
# ./kcli_cluster.sh create
parameters that contain master word need to be replaced with ctlplane. Found masters
```
so had to change this to `-P ctlplanes="$CLUSTER_MASTERS" \` Does that work on your version and if so can we also update the name of `CLUSTER_MASTERS` to be more inclusive. Thanks!

wainersm replied:
hi Steve!

I'm using a slightly older version of kcli: version: 99.0 commit: 8eeddd2 2022/10/03 Available Updates: N/A

The `ctlplanes` parameter worked out with that version. I will update the script to also remove `CLUSTER_MASTERS`

wainersm replied:
done

stevenhorsman replied:
Thanks - the changes look good

stevenhorsman commented on libvirt/README.md:
````diff
@@ -44,3 +44,222 @@ If you want to set default password for podvm debugging then you can use guestfi
+$ ssh-keygen -f ./id_rsa -N ""
+$ cat id_rsa.pub >> ~/.ssh/authorized_keys
+```
+**Note**: ensure that `~/.ssh/authorized_keys` has the right permissions (read/write for the user only) otherwise the
````
Is it worth suggest they do `chmod 600 ~/.ssh/authorized_keys` here to ensure this?

wainersm replied:
Done.

stevenhorsman commented on libvirt/README.md:
````diff
@@ -44,3 +44,222 @@ If you want to set default password for podvm debugging then you can use guestfi
+```
+$ cd ../podvm
+$ docker build --no-cache -t podvm_builder -f Dockerfile.podvm_builder .
+$ docker build --no-cache -t podvm_libvirt --build-arg BUILDER_IMG=podvm_builder:latest \
````
When running this I got an error in the libsecomp stage:
```
checking whether the C compiler works... no
configure: error: in `/tmp/build-libseccomp.Pv69G/libseccomp-2.5.4':
configure: error: C compiler cannot create executables
See `config.log' for more details
make: *** [Makefile.inc:100: /src/cloud-api-adaptor/podvm/staticlib/kata-libseccomp/lib/libseccomp.a] Error 77
```
I know it's not changed in the PR, but I thought it was worth noting.

wainersm replied:
The ci-e2e_libvirt branch on my git tree misses a fix that I sent yesterday in another PR: https://github.com/confidential-containers/cloud-api-adaptor/pull/537

Please rebase your tree and then those commands should work out.

wainersm replied:
Updated my tree.

stevenhorsman replied:
I hit a timeout in the image creation again which was quite upsetting after waiting ~1 hour for it to run:
```
==> qemu.ubuntu: Waiting for SSH to become available...
==> qemu.ubuntu: Timeout waiting for SSH.
==> qemu.ubuntu: Deleting output directory...
Build 'qemu.ubuntu' errored after 5 minutes 57 seconds: Timeout waiting for SSH.

==> Wait completed after 5 minutes 57 seconds

==> Some builds didn't complete successfully and had errors:
--> qemu.ubuntu: Timeout waiting for SSH.

==> Builds finished but no artifacts were created.
make: *** [Makefile:69: podvm-f080e93-dirty-amd64.qcow2] Error 1
```

I tried updating `qcow2/ubuntu/qemu-ubuntu.pkr.hcl` to have `ssh_wait_timeout  = "600s"`, but I'm not sure if that's the correct setting as it didn't seem to help and ` 5 minutes 57 seconds` is longer than the default of 300s?
```
==> qemu.ubuntu: Timeout waiting for SSH.
==> qemu.ubuntu: Deleting output directory...
Build 'qemu.ubuntu' errored after 6 minutes 31 seconds: Timeout waiting for SSH.

==> Wait completed after 6 minutes 31 seconds

==> Some builds didn't complete successfully and had errors:
--> qemu.ubuntu: Timeout waiting for SSH.

==> Builds finished but no artifacts were created.
make: *** [Makefile:69: podvm-f080e93-dirty-amd64.qcow2] Error 1
```

stevenhorsman replied:
It's worth noting that this passed for me in less than 4 mins when I ran it not in the container:
```
Build 'qemu.ubuntu' finished after 3 minutes 58 seconds.

==> Wait completed after 3 minutes 58 seconds
```

snir911 replied:
@wainersm, @stevenhorsman , i didn't hit this issue but one thing that may improve it is to pass `--device=/dev/kvm` (or even --privileged ) as it seems the VM starting inside the container is not accelerated with kvm, in any case worth mentioning it i think
Thanks to @whitebrandy who noticed this!

stevenhorsman replied:
@snir911  hey Snir, sorry for being slow, but where do you mean to add `--device=/dev/kvm`. It looks like `docker build`, doesn't take that as an argument, just `docker run`?
```
podvm# docker build --no-cache --device=/dev/kvm -t podvm_builder -f Dockerfile.podvm_builder .
unknown flag: --device
See 'docker build --help'.
```

snir911 replied:
oh, podman does, so i took it for granted also docker :P

stevenhorsman replied:
So good news and bad news on this I guess: I still can't get docker to work and the suggested flags don't exist, but I switched to podman (and added `--device=/dev/kvm`) and it all ran without an issue. 

I guess the question is what to do here:
- Change to documentation to require podman
- Add a note to say that if docker fails to try pdoman
- Try something else to get docker working

stevenhorsman replied:
As another data point - I took a fresh VM and running podman without `--device=/dev/kvm` also fails to build the image, so on my set-up (4 vCPU, 8GB VM) it seems like that is mandatory to get success here.

wainersm replied:
If you don't pass /dev/kvm then QEMU will run on emulated (TCG) mode, which is slower than hardware virtualization (KVM).  We cannot use KVM on cross-build nor on github jobs (e.g. https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/.github/workflows/podvm.yaml) because Github default runners don't support nested virt.

We can try to speed up the emulated mode. For example, use `q35` instead of `pc` machine type (https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/podvm/qcow2/ubuntu/variables.pkr.hcl#L56). And/or increase a little bit the requested memory (2GB) - https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/podvm/qcow2/ubuntu/variables.pkr.hcl#L31

On the podvm/README.md (see https://github.com/confidential-containers/cloud-api-adaptor/issues/550) document the environment constraints and required minimum memory/disk.

I don't like requiring podman. It seems more of our developers run docker.

Anyway, doesn't things shouldn't be addressed on this PR. @stevenhorsman @snir911 do you think it is good enough to get merged?

stevenhorsman replied:
So for this particular issue of getting the podvm built in a container, I'm ok for the tidy up to be added to part of #550.

For the wider PR, I'm a bit more mixed. - I think it's a great start, but the fact that I've not been able to get the scripts working and produce a libvirt peer pod set-up does bother me. It suggests that there is either a problem in the code (unlikely), a problem in what I'm doing (likely, but that indicates there is possibly a documentation problem), or a issue in my set-up that is different (again not something that is probably not documented). My concern with merging it so that we can then build on it is if we'll end up in a situation that no-one outside of RedHat can get it working, which isn't a great place. If it is just me that can't get it working, then I'm fine with it though.

On the other hand if we can merge then, then start automating and improving some of the steps e.g. with ansible scripts, and get a libvirt CI/CD running in the next step after this in open-source then that would be a huge step and worth 'breaking some eggs' for.

snir911 replied:
@stevenhorsman I'm pretty sure it would work for you with more resources, for sure if it runs on baremetal and not in a VM (as the VM packer uses is nested and in addition it's not accelerated) . Maybe we can have a note mentioning this intensive operation that requires a lot of resources and shouldn't run in a VM or something.
BTW when i tried to run the build for RHEL in a VM it failed due to disk size, I'm currently trying to optimize it on this aspect (https://github.com/confidential-containers/cloud-api-adaptor/pull/525) IDK if we can improve it on the other aspects as well

snir911 replied:
It seems emulating x86_64 on m1 has lower performanc,. I think it worth trying 8vCPUs / 32GB if it's not a burden

stevenhorsman replied:
FYI - the podvm build with docker worked fine in my 8 vCPU, 32GB RAM IBM Cloud VSI

wainersm replied:
I tried within a VM with 4 vCPUs, 8GB Mem, and 30GB disk on my Linux Fedora 34 laptop. It worked out.

Of course the build still requires much disk (being addressed in https://github.com/confidential-containers/cloud-api-adaptor/pull/525) and packer within container takes a good amount of time because QEMU emulated. This last problem can be mitigated by passing /dev/kvm (i.e. allowing KVM acceleration on container). I will be documenting all this on the PR for https://github.com/confidential-containers/cloud-api-adaptor/issues/550

whitebrandy replied:
> @wainersm, @stevenhorsman , i didn't hit this issue but one thing that may improve it is to pass `--device=/dev/kvm` (or even --privileged ) as it seems the VM starting inside the container is not accelerated with kvm, in any case worth mentioning it i think Thanks to @whitebrandy who noticed this!

So, back to the original comment, the closest I know of is specifying "RUN --security=insecure" which is similar to running "docker run --privileged" but sorry, untested :)

stevenhorsman commented on libvirt/README.md:
````diff
@@ -44,3 +44,222 @@ If you want to set default password for podvm debugging then you can use guestfi
+peer-pods-worker-0   Ready    worker                 2m47s   v1.25.3
+```
+
+## Prepare the Pod VM volume
````
Can we move this section up to the top please. At the moment there is a far bit of duplication between this and the **Creation of base storage volume** section and if you are like me then we will have already spent far too long finding and installing the correct version of the pre-reqs and building the podvm image before you see the option to do it in the container.

wainersm replied:
Talking about the podvm build in container, I opened an issue https://github.com/confidential-containers/cloud-api-adaptor/issues/550, it should be documented. I don't know if @bpradipt has already written about it somewhere, otherwise I can assign this task to myself.

I general i think libvirt's README should be re-write. So if you don't mind to merge this PR without major changes on it, then I can address the re-write in another PR.

stevenhorsman replied:
Yes - I think that's fine given that you've created the follow on issue for it. Thanks

stevenhorsman commented on libvirt/README.md:
```diff
@@ -44,3 +44,222 @@ If you want to set default password for podvm debugging then you can use guestfi
+$ cd ../../../libvirt/
+$ export LIBVIRT_IP="192.168.1.107"
+$ export SSH_KEY_FILE="id_rsa"
+$ ./install_operator.sh
```
When I ran this I hit an error:
```
# ./install_operator.sh
kubectl apply -f install/yamls/deploy.yaml
namespace/confidential-containers-system created
serviceaccount/cc-operator-controller-manager created
role.rbac.authorization.k8s.io/cc-operator-leader-election-role created
clusterrole.rbac.authorization.k8s.io/cc-operator-manager-role created
clusterrole.rbac.authorization.k8s.io/cc-operator-metrics-reader created
clusterrole.rbac.authorization.k8s.io/cc-operator-proxy-role created
rolebinding.rbac.authorization.k8s.io/cc-operator-leader-election-rolebinding created
clusterrolebinding.rbac.authorization.k8s.io/cc-operator-manager-rolebinding created
clusterrolebinding.rbac.authorization.k8s.io/cc-operator-proxy-rolebinding created
configmap/cc-operator-manager-config created
service/cc-operator-controller-manager-metrics-service created
deployment.apps/cc-operator-controller-manager created
customresourcedefinition.apiextensions.k8s.io/ccruntimes.confidentialcontainers.org created
kubectl apply -k install/overlays/libvirt
E0119 08:44:17.502494   56037 memcache.go:255] couldn't get resource list for confidentialcontainers.org/v1beta1: the server could not find the requested resource
configmap/peer-pods-cm-libvirt created
secret/auth-json-secret-libvirt created
secret/peer-pods-secret-libvirt created
secret/ssh-key-secret-libvirt created
daemonset.apps/cloud-api-adaptor-daemonset-libvirt created
error: resource mapping not found for name: "ccruntime-sample-libvirt" namespace: "confidential-containers-system" from "install/overlays/libvirt": no matches for kind "CcRuntime" in version "confidentialcontainers.org/v1beta1"
ensure CRDs are installed first
make: *** [Makefile:88: deploy] Error 1
```
and it finished very quickly, so I don't know if there should have been a wait in there or something? When I checked the pods they had been created:
```
# kubectl get pods -n confidential-containers-system
NAME                                             READY   STATUS    RESTARTS   AGE
cc-operator-controller-manager-fbb5dcf9d-w4jts   2/2     Running   0          59s
cloud-api-adaptor-daemonset-libvirt-j88xv        1/1     Running   0          56s
```
I re-ran it and it worked this time: `ccruntime.confidentialcontainers.org/ccruntime-sample-libvirt created`

stevenhorsman commented on libvirt/README.md:
````diff
@@ -44,3 +44,222 @@ If you want to set default password for podvm debugging then you can use guestfi
+$ kubectl wait --for=condition=Ready pod/busybox
+pod/busybox condition met
+```
+
````
Creating the pod failed for me:
```
Events:
  Type     Reason                  Age   From               Message
  ----     ------                  ----  ----               -------
  Normal   Scheduled               62s   default-scheduler  Successfully assigned default/busybox to peer-pods-worker-0
  Warning  FailedMount             62s   kubelet            MountVolume.SetUp failed for volume "kube-api-access-67gfb" : failed to sync configmap cache: timed out waiting for the condition
  Normal   AddedInterface          57s   multus             Add eth0 [10.244.1.4/24] from cbr0
  Warning  FailedCreatePodSandBox  2s    kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: remote hypervisor call failed: ttrpc: closed: unknown
```
It's probably user error, but I'm not sure how to resolve it.

stevenhorsman replied:
I got a different error on this set-up:
```
Events:
  Type     Reason                  Age               From               Message
  ----     ------                  ----              ----               -------
  Normal   Scheduled               107s              default-scheduler  Successfully assigned default/busybox to peer-pods-worker-0
  Normal   AddedInterface          102s              multus             Add eth0 [10.244.1.12/24] from cbr0
  Warning  FailedCreatePodSandBox  61s               kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: remote hypervisor call failed: ttrpc: closed: unknown
  Normal   AddedInterface          53s               multus             Add eth0 [10.244.1.13/24] from cbr0
  Normal   AddedInterface          41s               multus             Add eth0 [10.244.1.14/24] from cbr0
  Normal   AddedInterface          21s               multus             Add eth0 [10.244.1.15/24] from cbr0
  Normal   AddedInterface          12s               multus             Add eth0 [10.244.1.16/24] from cbr0
  Warning  FailedCreatePodSandBox  6s (x4 over 47s)  kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: failed to connect to remote hypervisor socket: dial unix /run/peerpod/hypervisor.sock: connect: connection refused: unknown
  Normal   AddedInterface          2s                multus             Add eth0 [10.244.1.17/24] from cbr0
```

One thing I'll note is that containerd doesn't seem to be installed, which I'd expect the pre-install section to do  when the operator is installed?
```
# containerd version

Command 'containerd' not found, but can be installed with:

apt install containerd
```

stevenhorsman replied:
On my big boy 8vCPU 32GB RAM the libvirt peer pod worked! (almost). The `kubectl wait --for=condition=Ready pod/busybox` failed as it took just over 30seconds to be created, but I think if we can document that you might need a really big machine then I'm happy with this. Thanks for the patience all!

stevenhorsman commented on libvirt/README.md:
````diff
@@ -44,3 +44,222 @@ If you want to set default password for podvm debugging then you can use guestfi
+5.17.12-100.fc34.x86_64
+$ kubectl exec pod/busybox -- uname -r
+5.4.0-131-generic
+```
````
I wonder if a clean-up/uninstall might be useful? That can be another PR though.

wainersm replied:
Done.

stevenhorsman replied:
Sorry, I wasn't particularly clear here. I was more thinking about the operator install, in case a user wanted to update (which I think would be an uninstall and then re-install, but if installing over the top works then this is fine)?

wainersm replied:
ah, ok. I never installed a new version over the top, so no idea it works (@bpradipt ?). If so then it is a good use test case for the operator tests.

I will add some words about uninstall/install the operator.

wainersm replied:
Done

stevenhorsman reviewed (commented):
I've re-tried the process and it seems to be better now, but I've still hit a few errors and I think there are some places that the doc could be a bit more helpful. If you can assist with the problems that would be great.

wainersm commented:
Added the section "Delete Confidential Containers and cloud-api-adaptor from the cluster" on the README.md. Delete the operator is not libvirt specific so maybe in another PR I could move it to somewhere else.

snir911 commented on libvirt/README.md:
````diff
@@ -44,3 +44,286 @@ If you want to set default password for podvm debugging then you can use guestfi
+$ ssh-keygen -f ./id_rsa -N ""
+$ cat id_rsa.pub >> ~/.ssh/authorized_keys
+```
+**Note**: ensure that `~/.ssh/authorized_keys` has the right permissions (read/write for the user only) otherwise the
````
comment can be removed now?

wainersm replied:
I didn't get. Why remove it? Mind to expand a little bit on your comment?

snir911 replied:
you do chmod no?

wainersm commented:
Updated the libvirt/README.md to replace 192.168.1.107 with 192.168.122.1.

stevenhorsman commented on libvirt/README.md:
```diff
@@ -44,3 +44,286 @@ If you want to set default password for podvm debugging then you can use guestfi
++-----------+---------+-----------+---------------------------------------+
+|  Cluster  |   Type  |    Plan   |                  Vms                  |
++-----------+---------+-----------+---------------------------------------+
+| peer-pods | generic | peer-pods | peer-pods-master-0,peer-pods-worker-0 |
```
```suggestion
| peer-pods | generic | peer-pods | peer-pods-ctlplane-0,peer-pods-worker-0 |
```

stevenhorsman commented on libvirt/README.md:
```diff
@@ -44,3 +44,286 @@ If you want to set default password for podvm debugging then you can use guestfi
+
+## Create the Kubernetes cluster
+
+Use the [`kcli_cluster.sh`](./kcli_cluster.sh) script to create a simple two VMs (one master and one worker) cluster
```
```suggestion
Use the [`kcli_cluster.sh`](./kcli_cluster.sh) script to create a simple two VMs (one control plane and one worker) cluster
```

stevenhorsman commented on libvirt/README.md:
```diff
@@ -44,3 +44,286 @@ If you want to set default password for podvm debugging then you can use guestfi
++-----------+---------+-----------+---------------------------------------+
+$ kubectl get nodes
+NAME                 STATUS   ROLES                  AGE     VERSION
+peer-pods-master-0   Ready    control-plane,master   6m8s    v1.25.3
```
```suggestion
peer-pods-ctlplane-0   Ready    control-plane,master   6m8s    v1.25.3
```

stevenhorsman commented on libvirt/README.md:
```diff
@@ -44,3 +44,286 @@ If you want to set default password for podvm debugging then you can use guestfi
+$ virsh -c qemu:///system list
+ Id   Name                 State
+------------------------------------
+ 3    peer-pods-master-0   running
+ 4    peer-pods-worker-0   running
```
```suggestion
 1    peer-pods-ctlplane-0   running
 2    peer-pods-worker-0     running
```

stevenhorsman commented on libvirt/README.md:
```diff
@@ -44,3 +44,286 @@ If you want to set default password for podvm debugging then you can use guestfi
+$ virsh -c qemu:///system list
+ Id   Name                     State
+----------------------------------------
+ 5    peer-pods-master-0       running
+ 6    peer-pods-worker-0       running
+ 7    podvm-busybox-88a70031   running
```
```suggestion
 1    peer-pods-ctlplane-0     running
 2    peer-pods-worker-0       running
 3    podvm-busybox-41e459c1   running
```

stevenhorsman commented on libvirt/README.md:
```diff
@@ -44,3 +44,286 @@ If you want to set default password for podvm debugging then you can use guestfi
+$ virsh -c qemu:///system list
+ Id   Name                 State
+------------------------------------
+ 5    peer-pods-master-0   running
+ 6    peer-pods-worker-0   running
```
```suggestion
 1    peer-pods-ctlplane-0     running
 2    peer-pods-worker-0       running
```

stevenhorsman reviewed (commented):
Mostly doc changes from master -> ctlplane

wainersm commented:
Updated the libvirt/README.md to remove occurrences of `master` where appropriated.

stevenhorsman reviewed (approved):
I've managed to get this process working, including the podvm build with docker. The secret to it was using an 8 vCPU 32GB RAM IBM Cloud VSI to give it enough resources. Big thanks to everyone's patience. I think it might be worth a statement about requiring significant compute resources, particularly on a VM, but big thanks for everyones patience and help with this.

For my future reference here is the steps that I took to get the pre-reqs installed and check the flow:
```
sudo apt-get update
echo "Install docker"
sudo apt-get install     ca-certificates     curl     gnupg     lsb-release
sudo mkdir -p /etc/apt/keyrings
curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /etc/apt/keyrings/docker.gpg
echo   "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu \
  $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
sudo apt-get update
sudo apt-get install docker-ce docker-ce-cli containerd.io docker-compose-plugin
echo "Checkout Wainer's code"
export GOPATH="${HOME}/go"
repo="github.com/wainersm/cc-cloud-api-adaptor"
repo_dir="${GOPATH}/src/${repo}"
mkdir -p $(dirname "${repo_dir}")
git clone -b ci-e2e_libvirt "https://${repo}.git" ${repo_dir}
pushd $repo_dir
cd podvm/
docker build --no-cache -t podvm_builder -f Dockerfile.podvm_builder .
docker build --no-cache -t podvm_libvirt --build-arg BUILDER_IMG=podvm_builder:latest --build-arg CLOUD_PROVIDER=libvirt -f Dockerfile.podvm .
podvm_container=$(docker create podvm_libvirt:latest /bin/bash)
file_name=$(docker export ${podvm_container} | tar -t | grep podvm)
docker rm ${podvm_container}
ls -al
echo "Install make"
sudo apt-get install -y build-essential
echo "Install go"
curl -OL https://go.dev/dl/go1.19.5.linux-amd64.tar.gz
rm -rf /usr/local/go && tar -C /usr/local -xzf go1.19.5.linux-amd64.tar.gz
export PATH=$PATH:/usr/local/go/bin
go version
echo "Install libvirt, kcli, configure storage pool"
sudo apt install qemu-kvm libvirt-daemon-system libvirt-clients bridge-utils virtinst virt-manager
sudo usermod -aG libvirt $USER
sudo usermod -aG kvm $USER
curl https://raw.githubusercontent.com/karmab/kcli/main/install.sh | sudo bash
sudo kcli create pool -p /var/lib/libvirt/images default
sudo setfacl -m u:$(id -un):rwx /var/lib/libvirt/images
echo "Install kubectl"
curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl"
sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl
cd ../libvirt/
ssh-keygen -t rsa -C "stevenh-libvirt-pp"
./kcli_cluster.sh create
export KUBECONFIG=$HOME/.kcli/clusters/peer-pods/auth/kubeconfig
kcli list kube
kubectl get nodes
export IMAGE=$pwd../podvm/podvm.qcow2
ls -al $IMAGE
virsh vol-create-as --pool default --name podvm-base.qcow2 --capacity 20G --allocation 2G --prealloc-metadata --format qcow2
virsh vol-upload --vol podvm-base.qcow2 $IMAGE --pool default --sparse
virsh -c qemu:///system vol-info --pool default podvm-base.qcow2
cd ../install/overlays/libvirt
ssh-keygen -f ./id_rsa -N ""
cat id_rsa.pub >> ~/.ssh/authorized_keys
chmod 600 ~/.ssh/authorized_keys
virsh -c "qemu+ssh://$USER@192.168.122.1/system?keyfile=$(pwd)/id_rsa" nodeinfo
cd ../../../libvirt/
export LIBVIRT_IP="192.168.122.1"
export SSH_KEY_FILE="id_rsa"
./install_operator.sh
kubectl get pods -n confidential-containers-system
kubectl logs pod/cloud-api-adaptor-daemonset-libvirt-pdz8d  -n confidential-containers-system
kubectl get runtimeclass
virsh -c qemu:///system list
vim sample_busybox.yaml
kubectl apply -f sample_busybox.yaml
kubectl wait --for=condition=Ready pod/busybox
kubectl get pods
virsh -c qemu:///system list
uname -r
kubectl exec pod/busybox -- uname -r
kubectl delete -f sample_busybox.yaml
virsh -c qemu:///system list
kubectl get pods -n confidential-containers-system
cd ../
CLOUD_PROVIDER=libvirt make delete
kubectl get pods -n confidential-containers-system
kubectl delete -f install/yamls/deploy.yaml
kubectl get pods -n confidential-containers-system
```

wainersm commented:
Thanks @stevenhorsman and @snir911 for the nice reviews and comments!

I opened https://github.com/confidential-containers/cloud-api-adaptor/pull/568 to document the podvm build process, including some things we discussed/realized on this PR. Let me know if I missed something.
